### PR TITLE
mailto:-links without address should work as "share" #716

### PR DESCRIPTION
**Pull request by Boehrsi**
Fix for #716.

- Subject and body are now handled
- To avoid bigger changes the created share text is added as extra to the intent, as this is already handled by the app
- Tested for existing and new contacts
- Should also work for a combination of mail + subject + body

### CODE REVIEW

File: src/org/thoughtcrime/securesms/NewConversationActivity.java [hunk 1]
```diff
@@ -121,8 +125,6 @@ public void onClick(DialogInterface dialog, int which) {
   }
 
   private void openConversation(int chatId) {
-    final DcContext dcContext = DcHelper.getContext(this);
```
Comment (Boehrsi) [inline src/org/thoughtcrime/securesms/NewConversationActivity.java]: Wasn't used anymore, so cleaned up that one.

File: src/org/thoughtcrime/securesms/NewConversationActivity.java [hunk 2]
```diff
@@ -103,16 +110,13 @@ public void onContactSelected(int specialId, String addr) {
...
-                .setPositiveButton(android.R.string.ok, new DialogInterface.OnClickListener() {
-                  @Override
-                  public void onClick(DialogInterface dialog, int which) {
-                    int contactId = dcContext.createContact(null, addr);
-                    if(contactId==0) {
-                      Toast.makeText(NewConversationActivity.this, R.string.bad_email_address, Toast.LENGTH_LONG).show();
-                      return;
-                    }
-                    openConversation(dcContext.createChatByContactId(contactId));
+                .setPositiveButton(android.R.string.ok, (dialog, which) -> {
```
Comment (Boehrsi) [inline src/org/thoughtcrime/securesms/NewConversationActivity.java]: More lambdas = less code

**Comment by r10s** [comment]
@Boehrsi thanks for pushing this forward.

i've just tested it, for the testlink from the issue, it worked - but the shave-via-mail-links from golem do not. do you have an idea what the problem is there?

also, i would prefer to see the chatlist and not the contactlist to select a chat. can we change this? then, this share-functionality would be on pair with the normal android share-function and the forward-function. also, this way, things can be shared to groups.

**Comment by Boehrsi** [comment]
I will have a look regarding the Golem links. The change from contact list to chat list is a bit more work, as I have to redirect the intent to another activity then. I will also have a look into that one.

**Comment by Boehrsi** [comment]
Tested and working now with

- [Golem Link](mailto:?subject=Intel: Apple soll ab 2020 auf ARM-Prozessoren umschwenken&body=Artikel auf Golem.de lesen:%0Ahttps://glm.io/139561?m%0A)
- [Our test Link](mailto:?subject=foo&body=bar)
- [Mail to with recipient](mailto:test@testing.de)

File: src/org/thoughtcrime/securesms/NewConversationActivity.java [hunk 3]
```diff
@@ -43,6 +46,10 @@
   @SuppressWarnings("unused")
   private static final String TAG = NewConversationActivity.class.getSimpleName();
   private static final String MAILTO = "mailto";
+  private static final String SUBJECT = "subject";
```
Comment (Boehrsi) [inline src/org/thoughtcrime/securesms/NewConversationActivity.java]: Constants are good

File: src/org/thoughtcrime/securesms/NewConversationActivity.java [hunk 4]
```diff
@@ -61,22 +68,23 @@ private void handleIntent() {
...
-            }
-            if (textToShare != null && !textToShare.isEmpty()) {
-              getIntent().putExtra(ConversationActivity.TEXT_EXTRA, textToShare);
-            }
             String recipientsList = mailto.getTo();
             if(recipientsList != null && !recipientsList.isEmpty()) {
               String[] recipientsArray = recipientsList.split(",");
               if (recipientsArray.length >= 1) {
                 String recipient = recipientsArray[0];
+                if (textToShare != null && !textToShare.isEmpty()) {
```
Comment (Boehrsi) [inline src/org/thoughtcrime/securesms/NewConversationActivity.java]: If we have a recipient and also content, append it to the new chat.

File: src/org/thoughtcrime/securesms/NewConversationActivity.java [hunk 5]
```diff
@@ -61,22 +68,23 @@ private void handleIntent() {
...
+                  getIntent().putExtra(ConversationActivity.TEXT_EXTRA, textToShare);
+                }
                 onContactSelected(DcContact.DC_CONTACT_ID_NEW_CONTACT, recipient);
               }
+            } else {
+              Intent shareIntent = new Intent(this, ShareActivity.class);
```
Comment (Boehrsi) [inline src/org/thoughtcrime/securesms/NewConversationActivity.java]: If no recipient is given delegate the text to the ShareActivity

File: src/org/thoughtcrime/securesms/NewConversationActivity.java [hunk 6]
```diff
@@ -87,6 +95,29 @@ private void handleIntent() {
...
+    Map<String, String> mailtoQueryMap = getMailtoQueryMap(uri);
+    String textToShare = mailtoQueryMap.get(SUBJECT);
+    String body = mailtoQueryMap.get(BODY);
+    if (body != null && !body.isEmpty()) {
+      textToShare += "\n" + body;
+    }
+    return textToShare;
+  }
+
+  private Map<String, String> getMailtoQueryMap(Uri uri) {
```
Comment (Boehrsi) [inline src/org/thoughtcrime/securesms/NewConversationActivity.java]: Manual query "parsing" and "handling" as the MailTo stuff given by the systems fails ...

**Comment by Boehrsi** [comment]
@r10s added the delegation to the ShareActivity if no recipient is given, if one is given the old way is used.

**Comment by r10s** [comment]
cool, i will check this out :)

**Comment by r10s** [comment]
works in my tests now :+1:

**Pull request merged**